### PR TITLE
daml-lf: test equality and hashing of values

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR advance the state of Generic Maps #2256.

Here we add extensive test for equality and hashing of values.  

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [X] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags, if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/spec/daml-lf-1.rst [hunk 1]
```diff
@@ -2929,6 +3040,15 @@ comments::
   // * must be non empty *
 
 
+Maps
+....
+
+The program serialization format does not provide any direct way to
+encode neither `TextMap` not `GenMap`. DAML-LF programs can create
```
Comment (SamirTalwar) [inline daml-lf/spec/daml-lf-1.rst]:
```suggestion
encode either `TextMap` or `GenMap`. DAML-LF programs can create
```

File: daml-lf/spec/daml-lf-1.rst [hunk 2]
```diff
@@ -2564,8 +2728,11 @@ List functions
   predicate give as first argument.
 
 
-TextMap functions
-~~~~~~~~~~~~~~~~~
+Text map functions
+~~~~~~~~~~~~~~~~~~
+
+**Entries ordering**: All the operation above preserved the order of
+ the map according the lexicographic order of its keys.
```
Comment (SamirTalwar) [inline daml-lf/spec/daml-lf-1.rst]: Typos:

```rst
**Entry order**: The operations above preserve the order of
the map according to the lexicographic order of its keys.
```

Comment (SamirTalwar) [inline daml-lf/spec/daml-lf-1.rst]: I'm not sure what it means to both "preserve" the order and also use lexicographic ordering. (I'm probably missing something.) If it's ordered lexicographically, then it's sorted, and it's not necessary to "preserve" anything; there's an invariant in place.

File: daml-lf/spec/daml-lf-1.rst [hunk 3]
```diff
@@ -1686,7 +1825,7 @@ exact output.
 
   Evaluation result
     r ::= Ok v                                      -- ResOk
-       |  Err LitText                               -- ResErr
+       |  Err t                                    -- ResErr
```
Comment (SamirTalwar) [inline daml-lf/spec/daml-lf-1.rst]: Either add a space here or remove from the previous line.

```suggestion
       |  Err t                                     -- ResErr
```

**Comment by remyhaemmerle-da** [comment]
> Can't comment on the overall PR, as I don't understand large parts of it, but I noted a couple of typos.

Sorry, there were a mistake with the PR. I update the it. You can take a new look.

(Thanks for the typos, I will fixed that in another PR. )

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Hasher.scala [hunk 4]
```diff
@@ -69,15 +67,15 @@ private[speedy] object Hasher {
...
-                loop(pushOrderedValues(values.iterator().asScala, cmdsRest), stack)
+                loop(pushOrderedValues(values.size, values.iterator().asScala, cmdsRest), stack)
               case SVariant(_, variant, value) =>
                 loop(Value(value) :: Mix(variant.hashCode) :: cmdsRest, stack)
               case SStruct(_, values) =>
-                loop(pushOrderedValues(values.iterator().asScala, cmdsRest), stack)
+                loop(pushOrderedValues(values.size, values.iterator().asScala, cmdsRest), stack)
               case SOptional(opt) =>
-                loop(pushOrderedValues(opt.iterator, cmdsRest), stack)
+                loop(pushOrderedValues(opt.fold(0)(_ => 1), opt.iterator, cmdsRest), stack)
```
Comment (SamirTalwar) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Hasher.scala]: You can just use `opt.size` here.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Hasher.scala [hunk 5]
```diff
@@ -30,8 +28,8 @@ private[speedy] object Hasher {
   def hash(v: SValue): Int =
     loop(List(Value(v)))
 
-  private def pushOrderedValues(values: Iterator[SValue], cmds: List[Command]) =
-    ((Ordered(values.size) :: cmds) /: values) { case (acc, v) => Value(v) :: acc }
+  private def pushOrderedValues(size: Int, values: Iterator[SValue], cmds: List[Command]) =
+    ((Ordered(size) :: cmds) /: values) { case (acc, v) => Value(v) :: acc }
```
Comment (SamirTalwar) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Hasher.scala]: Nice catch.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Hasher.scala [hunk 6]
```diff
@@ -89,11 +87,11 @@ private[speedy] object Hasher {
                 }
                 loop(newCmds, stack)
               case SAny(t, v) =>
-                loop(Value(v) :: Mix(t.hashCode()) :: cmds, stack)
+                loop(Value(v) :: Mix(t.hashCode()) :: cmdsRest, stack)
```
Comment (SamirTalwar) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Hasher.scala]: I am glad you're writing tests.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/svalue/SEquatableValuesSpec.scala [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+import org.scalatest.{Matchers, WordSpec}
+
+import scala.collection.immutable.HashMap
+import scala.language.implicitConversions
+import java.util
+
+import com.digitalasset.daml.lf.language.{Ast, Util => AstUtil}
+import com.digitalasset.daml.lf.speedy.{SBuiltin, SExpr, SValue}
+
+@SuppressWarnings(Array("org.wartremover.warts.Any"))
```
Comment (SamirTalwar) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/svalue/SEquatableValuesSpec.scala]: Is this really necessary?

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/svalue/SEquatableValuesSpec.scala [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+      List(b, a, a),
+      List(a, b, c),
+      List(c, a, b),
+    )
+  }
+
+  private def optLists(atLeast3Values: List[SValue]) = {
+    val a :: b :: c :: _ = atLeast3Values
+    val x = SOptional(None)
+    List(
```
Comment (S11001001) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/svalue/SEquatableValuesSpec.scala]:
```scala
    val s = x :: (atLeast3Values take 3)
    List.empty :: (List(s).sequence ++ List(s, s).sequence ++ List(s, s, s).sequence)
```

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/svalue/SEquatableValuesSpec.scala [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+
+  private def mkStruct2(fst: List[SValue], snd: List[SValue]) =
+    for {
+      x <- fst
+      y <- snd
+    } yield SStruct(record2Fields, ArrayList(x, y))
+
+  private def lists(atLeast3Values: List[SValue]) = {
+    val a :: b :: c :: _ = atLeast3Values
+    List(
```
Comment (S11001001) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/svalue/SEquatableValuesSpec.scala]: Like `optLists` with `val s = List(a, b, c)`.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Hasher.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Hasher.scala]: implicit conversion from Option to Traversable are considered warts.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/svalue/SEquatableValuesSpec.scala [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+      forEvery(nonEquatableValues)(atoms =>
+        forEvery(atoms)(x => a[Hasher.NonHashableSValue] should be thrownBy Hasher.hash(x)))
+    }
+
+  }
+
+  private def ArrayList[X](as: X*): util.ArrayList[X] = {
+    val a = new util.ArrayList[X](as.length)
+    as.foreach(a.add)
+    a
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/svalue/SEquatableValuesSpec.scala]: fixed

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/svalue/SEquatableValuesSpec.scala (hunk 8 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/svalue/SEquatableValuesSpec.scala]: done.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/svalue/SEquatableValuesSpec.scala (hunk 9 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/svalue/SEquatableValuesSpec.scala]: done

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/svalue/SEquatableValuesSpec.scala (hunk 7 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/svalue/SEquatableValuesSpec.scala]: in fact  not. I dropped it.

**Comment by SamirTalwar** [comment]
I'm happy, and @S11001001 is offline for a few days.

**Pull request merged**